src/com/owncloud/android/services/observer/InstantUploadsHandler.java

```diff
@@ -0,0 +1,295 @@
+     * upload request will be still in the queue or being uploaded when the second one arrives.
+     *
+     * Nevertheless, when network is not available, the first upload request might be retired from the
+     * queue and archived as failed/delayed before the second upload request arrives to the service.
+     * That will lead to two failed uploads for the same file, and this will be uploaded twice when
+     * the network is recovered. Real case detected in
+     * https://github.com/owncloud/android/issues/1795#issuecomment-245263247.
+     *
+     * Even worse: due to multithreading scheduling, it's not granted that the two events corresponding
+     * to a single picture/video are received one after the other. If some pictures/videos are taken in a row,
+     * FileObserver thread can call 'handleNewMediaFile(...)' action for some of them before than
+     * InstantUploadsBroadcastReceiver calls 'handleNewXXXAction(...)' for the first of them. Real case
+     * detected while fixing https://github.com/owncloud/android/issues/1795#issuecomment-245263247.
+     *
+     * Due to that, remembering just the last file detected is not enough to correctly filter out duplicated
+     * detections. Next static field saves last recent requests to filter duplicated detections for a while.
+     */
+    private static int MAX_RECENTS = 30;
+    private static Set<String> sRecentlyUploadedFilePaths = new HashSet<>(MAX_RECENTS);
+
+
+    public boolean handleNewPictureAction(
+        Intent intent,
+        InstantUploadsConfiguration configuration,
+        Context context
+    ) {
+        Log_OC.i(TAG, "New photo received");
+
+        if (!configuration.isEnabledForPictures()) {
+            Log_OC.d(TAG, "Instant upload disabled for images, ignoring new picture");
+            return false;
+        }
+
+        /// retrieve file data from MediaStore
+        String[] CONTENT_PROJECTION = {
+            MediaStore.Images.Media.DATA,
+            MediaStore.Images.Media.DISPLAY_NAME,
+            MediaStore.Images.Media.MIME_TYPE,
+            MediaStore.Images.Media.SIZE
+        };
+
+        Cursor c = context.getContentResolver().query(intent.getData(), CONTENT_PROJECTION, null, null, null);
+        if (c == null || !c.moveToFirst()) {
+            Log_OC.e(TAG, "Couldn't resolve given uri: " + intent.getDataString());
+            if (c != null) {
+                c.close();
+            }
+            return false;
+        }
+        String localPath = c.getString(c.getColumnIndex(MediaStore.Images.Media.DATA));
+        String fileName = c.getString(c.getColumnIndex(MediaStore.Images.Media.DISPLAY_NAME));
+        String mimeType = c.getString(c.getColumnIndex(MediaStore.Images.Media.MIME_TYPE));
+        c.close();
+
+        Log_OC.d(TAG, "Local path: " + localPath);
+        return handleNewMediaFile(fileName, localPath, mimeType, true, configuration, context);
```

Comment:
Maybe the functions `handleNewPictureAction `and `handleNewVideoAction` could be merged in only one, since the media attributes are both from `MediaStore.MediaColumns` and the only difference would be the configuration option for instant upload enabled. For the latter, the intent action could be checked here instead of the broadcast receiver.